### PR TITLE
Single security input

### PR DESCRIPTION
**Pull request by spiresheep**

### CODE REVIEW

File: Applications/WebPortal/WebApp/library/source/components/index.ts [hunk 1]
```diff
@@ -9,4 +9,5 @@ export * from './integer_input_box';
 export * from './modal';
 export * from './money_input_box';
 export * from './number_input';
+export * from './single_security_input';
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/index.ts]: Just like you don't name it ```single_number_input```, similarly don't name it ```single_security_input```. The name ```security_input``` is sufficient.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 2]
```diff
@@ -0,0 +1,690 @@
...
+
+interface State {
+  inputString: string;
+  isEditing: boolean;
+  localValue: Nexus.Security;
+  isSelected: boolean;
+}
+
+/** A component that displays a list of securities. */
+export class SingleSecurityInput extends React.Component<Properties, State> {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Rename to ```SecurityInput```.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+  public static readonly defaultProps = {
+    value: '',
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: ```value``` is of type ```Security```.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+interface Properties {
+
+  /** The size at which the component should be displayed at. */
+  displaySize: DisplaySize;
+
+  /** Determines if the component is readonly. */
+  readonly?: boolean;
+
+  /** The list of securities to display. */
+  value?: Nexus.Security;
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: There is no list.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+
+  /** Called when the list of values changes.
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: There is no list.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+
+  public render() {
+    const visibility = (() => {
+      if(!this.state.isEditing) {
+        return SingleSecurityInput.STYLE.hidden;
+      } else {
+        return null;
+      }
+    })();
+    const inputField = (() => {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: No need for this to be within a function or even declared here. Inline this directly into the JSX below.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+    })();
+    const imageSize = (() => {
+      if(this.props.displaySize === DisplaySize.SMALL) {
+        return SingleSecurityInput.IMAGE_SIZE_SMALL_VIEWPORT;
+      } else {
+        return SingleSecurityInput.IMAGE_SIZE_LARGE_VIEWPORT;
+      }
+    })();
+    const iconWrapperStyle = (() => {
+      const displaySize = this.props.displaySize;
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Remove the ```displaySize```.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+      } else {
+        return SingleSecurityInput.STYLE.iconWrapperLarge;
+      }
+    })();
+    const removeButton = (() => {
+      if(this.state.isSelected) {
+        if(this.props.displaySize === DisplaySize.SMALL) {
+          return (
+            <div style={iconWrapperStyle}
+              onClick={this.removeEntry}>
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Fix indentation.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+              <img height={imageSize} width={imageSize}
+                style={SingleSecurityInput.STYLE.iconClickableStyle}
+                src={SingleSecurityInput.PATH + 'remove-purple.svg'}/>
+            </div>);
+        } else {
+          return (
+            <div style={iconWrapperStyle}
+              onClick={this.removeEntry}>
+              <img height={imageSize} width={imageSize}
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Fix indentation.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+        if(this.props.displaySize === DisplaySize.SMALL) {
+          return (
+            <div style={iconWrapperStyle}>
+              <img height={imageSize} width={imageSize}
+                src={SingleSecurityInput.PATH + 'remove-grey.svg'}/>
+            </div>);
+        } else {
+          return (
+            <div style={iconWrapperStyle}
+              onClick={this.removeEntry}>
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Indentation.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+      <div>
+        <input
+          style={SingleSecurityInput.STYLE.textBox}
+          className={css(SingleSecurityInput.EXTRA_STYLE.effects)}
+          value={displayValue}
+          onFocus={this.onOpen}
+          onClick={this.onOpen}/>
+        <div style={visibility}>
+          <Modal displaySize={this.props.displaySize}
+            width='300px' height={SingleSecurityInput.MODAL_HEIGHT}
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Fix indentation.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+  private removeEntry() {
+    if(this.state.isSelected) {
+      this.setState({
+        isSelected: false,
+        localValue: null
+      });
+    }
+  }
+
+  private selectEntry(bool: boolean) {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Don't use ```bool``` as a variable name. Rename to ```isSelected``` or some other variant of ```isXXX```.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 13]
```diff
@@ -0,0 +1,690 @@
...
+    this.setState({ isSelected: bool});
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Fix spacing.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 14]
```diff
@@ -0,0 +1,690 @@
...
+  private static readonly IMAGE_SIZE_LARGE_VIEWPORT = '16px';
+  private static readonly MODAL_HEADER = 'Edit Symbols';
+  private static readonly MODAL_HEIGHT = '363px';
+  private static readonly PATH =
+    'resources/account_page/compliance_page/security_input/';
+  private static readonly REMOVE_TEXT = 'Remove';
+  private static readonly SUBMIT_CHANGES_TEXT = 'Submit Changes';
+}
+
+interface InputFieldProperties {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: If this is the same as ```SecurityListInput```, then factor it out into its own file and import it.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx [hunk 15]
```diff
@@ -0,0 +1,690 @@
...
+      '::moz-focus-inner': {
+        border: 0
+      }
+    }
+  });
+  private static readonly PLACEHOLDER_TEXT = 'Find symbol here';
+}
+
+
+interface SymbolsBoxProperties {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: ```SymbolBoxProperties```.

File: Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx (hunk 3 above)
Comment (spiresheep) [reply Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: Is there a Nexus.Security.None type variable I can use? Or should I make it null?

Comment (kranar) [reply Applications/WebPortal/WebApp/library/source/components/single_security_input.tsx]: No, don't make it ```null```. Instead add a static value called ```NONE``` to Nexus/WebApi/source/definitions/security.ts, the symbol should be ```''```, the market is ```MarketCode.NONE``` and the country is ```CountryCode.NONE```.

File: Applications/WebPortal/WebApp/library/source/components/securities_field.tsx [hunk 16]
```diff
@@ -1,7 +1,7 @@
 import { css, StyleSheet } from 'aphrodite';
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_field.tsx]: Rename ```single_security_input.tsx``` to ```security_input.tsx```

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 17]
```diff
@@ -0,0 +1,673 @@
...
+    })();
+    let displayValue  = '';
+    for(let i = 0; i < this.props.value.length; ++i) {
+      const symbol = this.props.value[i].symbol;
+      displayValue = displayValue.concat(symbol);
+      if(this.props.value.length > 1 && i < this.props.value.length - 1) {
+        displayValue = displayValue.concat(', ');
+      }
+    }
+    return(
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: Space after ```return```.

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 18]
```diff
@@ -0,0 +1,673 @@
...
+  inputString: string;
+  isEditing: boolean;
+  localValue: Nexus.Security[];
+  selection: number;
+}
+
+/** A component that displays a list of securities. */
+export class SecuritiesInput extends React.Component<Properties, State> {
+  public static readonly defaultProps = {
+    value: '',
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: This is not an acceptable default ```value```.

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 19]
```diff
@@ -0,0 +1,673 @@
...
+        return SecuritiesInput.MODAL_HEADER_READONLY;
+      } else {
+        return SecuritiesInput.MODAL_HEADER;
+      }
+    })();
+    const inputField = (() => {
+      if(this.props.readonly) {
+        return null;
+      } else {
+        return(
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: Space after ```return```.

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 20]
```diff
@@ -0,0 +1,673 @@
...
+    const imageSize = (() => {
+      if(this.props.displaySize === DisplaySize.SMALL) {
+        return SecuritiesInput.IMAGE_SIZE_SMALL_VIEWPORT;
+      } else {
+        return SecuritiesInput.IMAGE_SIZE_LARGE_VIEWPORT;
+      }
+    })();
+    const iconWrapperStyle = (() => {
+      const displaySize = this.props.displaySize;
+      if(displaySize=== DisplaySize.SMALL && this.props.readonly) {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: Space before ```===```

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 21]
```diff
@@ -0,0 +1,673 @@
...
+  onClick?: (index: number) => void;
+}
+
+/** A component that displays a list of symbols. */
+class SymbolsField extends React.Component<SymbolsFieldProperties> {
+  public render() {
+    const scrollHeader = (() => {
+      if(!this.props.readonly) {
+        if(this.props.displaySize === DisplaySize.SMALL) {
+          return(
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: Space after ```return```.

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 22]
```diff
@@ -0,0 +1,673 @@
...
+            <div style={SymbolsField.STYLE.scrollBoxHeaderSmall}>
+              {'Added Symbols'}
+            </div>);
+        } else {
+          return(
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: See above.

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 23]
```diff
@@ -0,0 +1,673 @@
...
+
+/** A component that displays a list of symbols. */
+class SymbolsField extends React.Component<SymbolsFieldProperties> {
+  public render() {
+    const scrollHeader = (() => {
+      if(!this.props.readonly) {
+        if(this.props.displaySize === DisplaySize.SMALL) {
+          return(
+            <div style={SymbolsField.STYLE.scrollBoxHeaderSmall}>
+              {'Added Symbols'}
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: Don't put string literals within curly brackets.  ```<div>Added Symbols</div>``` is fine.

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 24]
```diff
@@ -0,0 +1,673 @@
...
+            </div>);
+        } else {
+          return(
+            <div style={SymbolsField.STYLE.scrollBoxHeaderLarge}>
+              {'Added Symbols'}
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: See above.

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 25]
```diff
@@ -0,0 +1,673 @@
...
+        }
+      } else {
+        return null;
+      }
+    })();
+    const selectedSecuritiesBox = (() => {
+      const displaySize = this.props.displaySize;
+      if(displaySize === DisplaySize.SMALL && this.props.readonly) {
+        return SymbolsField.STYLE.scrollBoxSmallReadonly;
+      } else if (displaySize === DisplaySize.SMALL && !this.props.readonly ) {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: No space after ```if``` or before the closing ```)```.

File: Applications/WebPortal/WebApp/library/source/components/securities_input.tsx [hunk 26]
```diff
@@ -0,0 +1,673 @@
...
+        return SymbolsField.STYLE.scrollBoxSmall;
+      } else if(displaySize !== DisplaySize.SMALL && this.props.readonly) {
+        return SymbolsField.STYLE.scrollBoxBigReadonly;
+      } else if (displaySize !== DisplaySize.SMALL && !this.props.readonly) {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/securities_input.tsx]: No space after the ```if```

File: Applications/WebPortal/WebApp/library/source/components/security_input.tsx [hunk 27]
```diff
@@ -451,242 +398,93 @@ export class SecurityInput extends React.Component<Properties, State> {
...
-            <div style={SymbolsBox.STYLE.scrollBoxHeaderLarge}>
-              {'Added Symbols'}
-            </div>);
-        }
-      } else {
-        return null;
+    const displayValue = (() => {
+      if(this.props.value !== null) {
+        return this.props.value.symbol;
+      }else {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/security_input.tsx]: Fix spacing.

File: Applications/WebPortal/WebApp/library/source/components/security_input.tsx [hunk 28]
```diff
@@ -451,242 +398,93 @@ export class SecurityInput extends React.Component<Properties, State> {
...
-      if(displaySize === DisplaySize.SMALL && this.props.readonly) {
-        return SymbolsBox.STYLE.scrollBoxSmallReadonly;
-      } else if (displaySize === DisplaySize.SMALL && !this.props.readonly ) {
-        return SymbolsBox.STYLE.scrollBoxSmall;
-      } else if(displaySize !== DisplaySize.SMALL && this.props.readonly) {
-        return SymbolsBox.STYLE.scrollBoxBigReadonly;
-      } else if (displaySize !== DisplaySize.SMALL && !this.props.readonly) {
-        return SymbolsBox.STYLE.scrollBoxBig;
+    const isValidSecurity = (() => {
+      if(this.props.value === null) {
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/security_input.tsx]: ```const isValidSecurity = this.props.value !== null && this.props.value.symbol !== '';```

File: Applications/WebPortal/WebApp/library/source/components/security_input.tsx [hunk 29]
```diff
@@ -451,242 +398,93 @@ export class SecurityInput extends React.Component<Properties, State> {
...
-            </div>);
-        }
-    }
     return (
-      <div style={selectedSecuritiesBox}>
-        {scrollHeader}
-        {entries}
+      <div style={SymbolField.STYLE.scrollBoxSmall}>
+        <div style={SymbolField.STYLE.scrollBoxHeader}>
+          {'Added Symbol'}
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/security_input.tsx]: No curly brackets or quotes.

File: Applications/WebPortal/WebApp/library/source/components/security_input.tsx [hunk 30]
```diff
@@ -451,242 +398,93 @@ export class SecurityInput extends React.Component<Properties, State> {
...
+        </div>
+        <div className={css(entryStyle)}
+          onClick={this.onClick}>
```
Comment (kranar) [inline Applications/WebPortal/WebApp/library/source/components/security_input.tsx]: Double indent.

**Pull request merged**